### PR TITLE
[FLINK-5047] [table] Add sliding group-windows for batch tables

### PR DESCRIPTION
**Pull request by twalthr**
This PR implements sliding group-windows. It covers the following cases:

- Grouped Slide count-windows with incremental aggregates
- Grouped and All Slide event-time time-windows with incremental and non-incremental aggregates

All windows support the overlapping and non-overlapping case. Slide windows are pre-tumbled if possible. This PR also fixes some bugs. 

If the general design is ok, I will also implement the missing non-incremental aggregates for count-windows. And add a bit of documentation.

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideCountWindowAggReduceGroupFunction.scala [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    private val windowSlide: Long,
+    @transient private val returnType: TypeInformation[Row])
+  extends RichGroupReduceFunction[Row, Row]
+  with ResultTypeQueryable[Row] {
+
+  private var output: Row = _
+  private var outWindowStartIndex: Int = _
+
+  override def open(config: Configuration) {
+    Preconditions.checkNotNull(aggregates)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideCountWindowAggReduceGroupFunction.scala]: I think we can move everything into the constructor and remove the `open()` method.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideCountWindowAggReduceGroupFunction.scala [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        aggregates.foreach(_.initiate(output))
+      }
+
+      // merge intermediate aggregate value to buffer.
+      aggregates.foreach(_.merge(record, output))
+
+      count += 1
+
+      // trigger tumbling evaluation
+      if (count % preTumblingSize == 0) {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideCountWindowAggReduceGroupFunction.scala]: do we also have to emit "incomplete" windows after the `while (iterator.hasNext)` loop?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetWindowAggregate.scala [hunk 3]
```diff
@@ -280,6 +285,138 @@ class DataSetWindowAggregate(
...
+            namedAggregates,
+            grouping,
+            inputType,
+            isParserCaseSensitive)
+
+          mappedDataSet.asInstanceOf[DataSet[Row]]
+            .groupBy(groupingKeys: _*)
+            // sort on time field, it's the last element in the row
+            .sortGroup(mapReturnType.getArity - 1, Order.ASCENDING)
+            .reduceGroup(prepareReduceFunction) // pre-tumbles and replicates/omits
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetWindowAggregate.scala]: Only do this if the tumble size is > 1?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggFlatMapFunction.scala [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    private val windowSlide: Long,
+    @transient private val returnType: TypeInformation[Row])
+  extends RichFlatMapFunction[Row, Row]
+  with ResultTypeQueryable[Row] {
+
+  private var aggregateBuffer: Row = _
+  private var outWindowStartIndex: Int = _
+
+  override def open(config: Configuration) {
+    Preconditions.checkNotNull(aggregates)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggFlatMapFunction.scala]: Move everything to the constructor and remove `open()`?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggReduceCombineFunction.scala [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+        aggregateBuffer.setField(timeFieldPos, record.getField(timeFieldPos))
+
+        return aggregateBuffer
+      }
+    }
+
+    // this code path should never be reached as we return before the loop finishes
+    throw new IllegalArgumentException("Group is empty. This should never happen.")
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggReduceCombineFunction.scala]: can be removed. It's the responsibility of the DataSet API to call the user functions correctly. And even if it would not, this function would behave correctly.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggReduceGroupFunction.scala [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    private val windowSlide: Long,
+    @transient private val returnType: TypeInformation[Row])
+  extends RichGroupReduceFunction[Row, Row]
+  with ResultTypeQueryable[Row] {
+
+  protected var aggregateBuffer: Row = _
+  private var outWindowStartIndex: Int = _
+
+  override def open(config: Configuration) {
+    Preconditions.checkNotNull(aggregates)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggReduceGroupFunction.scala]: Move checks and initialization to constructor and remove `open()`?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideWindowAggReduceGroupFunction.scala [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+  extends RichGroupReduceFunction[Row, Row] {
+
+  protected var aggregateBuffer: Row = _
+  protected var windowStartFieldPos: Int = _
+
+  private var collector: TimeWindowPropertyCollector = _
+  private var output: Row = _
+
+  override def open(config: Configuration) {
+    Preconditions.checkNotNull(aggregates)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideWindowAggReduceGroupFunction.scala]: Except for the initialization of `TimeWindowPropertyCollector` everything can be moved into the constructor.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/batch/table/GroupWindowTest.scala [hunk 8]
```diff
@@ -360,6 +360,19 @@ class GroupWindowTest extends TableTestBase {
       .window(Session withGap 7.milli as 'w) // require on a time attribute
       .groupBy('string, 'w)
       .select('string, 'int.count)
+
+    val expected = unaryNode(
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/batch/table/GroupWindowTest.scala]: Why do you add the expected result to a test which is expected to fail?

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/table/AggregationsITCase.scala [hunk 9]
```diff
@@ -146,42 +146,9 @@ class AggregationsITCase extends StreamingMultipleProgramsTestBase {
       "Hi,1,1,1970-01-01 00:00:00.0,1970-01-01 00:00:00.005")
     assertEquals(expected.sorted, StreamITCase.testResults.sorted)
   }
-
-  @Test
-  def testEventTimeSlidingWindow(): Unit = {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/table/AggregationsITCase.scala]: Is this test subsumed by `DataStreamAggregateITCase`?

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/dataset/DataSetWindowAggregateITCase.scala [hunk 10]
```diff
@@ -169,4 +168,228 @@ class DataSetWindowAggregateITCase(configMode: TableConfigMode)
...
+
+    val table = env.fromCollection(data).toTable(tEnv, 'long, 'int, 'string)
+
+    val windowedTable = table
+      .window(Slide over 5.milli every 10.milli on 'long as 'w)
+      .groupBy('string, 'w)
+      .select('string, 'int.count, 'w.start, 'w.end)
+
+    val expected =
+      "Hallo,1,1970-01-01 00:00:00.0,1970-01-01 00:00:00.005\n" +
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/dataset/DataSetWindowAggregateITCase.scala]: There are two result rows for the same window (Hallo, 00:00:00.0, 00:00:00.005). Is this correct?

**Comment by twalthr** [comment]
@fhueske thanks for your comments. I will create a time window PR.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala [hunk 11]
```diff
@@ -153,6 +170,156 @@ object AggregateUtil {
...
+          // for incremental aggregations
+          new DataSetSlideTimeWindowAggReduceCombineFunction(
+            aggregates,
+            groupings.length,
+            returnType.getArity - 1,
+            asLong(size),
+            asLong(slide),
+            returnType)
+        } else {
+          // for non-incremental aggregations
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: You are right. I will remove this case distinction.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggFlatMapFunction.scala (hunk 4 above)
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggFlatMapFunction.scala]: Row is not always serializable and accumulators too. We are safer if we init them in `open()`.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggFlatMapFunction.scala [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    val windowStart = record.getField(timeFieldPos).asInstanceOf[Long]
+
+    // adopted from SlidingEventTimeWindows.assignWindows
+    var start: Long = TimeWindow.getWindowStartWithOffset(windowStart, 0, windowSlide)
+
+    // skip preparing output if it is not necessary
+    if (start > windowStart - windowSize) {
+
+      // prepare output
+      for (i <- aggregates.indices) {
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggFlatMapFunction.scala]: You are right. Will simplify that.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideCountWindowAggReduceGroupFunction.scala [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+  * for different panes of a sliding window.
+  *
+  * @param aggregates aggregate functions
+  * @param groupingKeysLength number of grouping keys
+  * @param preTumblingSize number of records to be aggregated (tumbled) before emission
+  * @param windowSize window size of the sliding window
+  * @param windowSlide window slide of the sliding window
+  * @param returnType return type of this function
+  */
+class DataSetSlideCountWindowAggReduceGroupFunction(
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideCountWindowAggReduceGroupFunction.scala]: I will create an issue for this.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideCountWindowAggReduceGroupFunction.scala (hunk 2 above)
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideCountWindowAggReduceGroupFunction.scala]: I don't think so. It returns the same result than the DataStream variant.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggFlatMapFunction.scala (hunk 4 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggFlatMapFunction.scala]: An empty `Row` can be serialized because all fields are null and it implements `Serializable`. However, once we have a non-serializable value in the row, it will fail. 
But I think you are right, lets keep the `Row` initialization in `open()`.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideWindowAggReduceCombineFunction.scala [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        aggregateBuffer.setField(windowStartFieldPos, record.getField(windowStartFieldPos))
+
+        return aggregateBuffer
+      }
+    }
+
+    // this code path should never be reached as we return before the loop finishes
+    throw new IllegalArgumentException("Group is empty. This should never happen.")
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideWindowAggReduceCombineFunction.scala]: I cannot remove it. The compiler complains otherwise.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/table/AggregationsITCase.scala (hunk 9 above)
Comment (twalthr) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/table/AggregationsITCase.scala]: Yes.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/batch/table/GroupWindowTest.scala (hunk 8 above)
Comment (twalthr) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/batch/table/GroupWindowTest.scala]: Seems that I haven't seen the exception. Thanks.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideCountWindowAggReduceGroupFunction.scala (hunk 2 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideCountWindowAggReduceGroupFunction.scala]: But shouldn't batch be the reference for streaming?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideWindowAggReduceCombineFunction.scala (hunk 14 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideWindowAggReduceCombineFunction.scala]: OK, but how about we make that more clear with a comment like:

`// This will never happen because the iterator is never null but we have to satisfy the compiler.`

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideCountWindowAggReduceGroupFunction.scala (hunk 2 above)
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideCountWindowAggReduceGroupFunction.scala]: I thought about this again. I think it is ok not to emit incomplete tumbling windows. If the count is not reached in the tumbling window it does not trigger a sliding window.

**Comment by twalthr** [comment]
Thanks for the feedback @fhueske. I updated the PR.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggFlatMapFunction.scala [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    private val timeFieldPos: Int,
+    private val windowSize: Long,
+    private val windowSlide: Long,
+    @transient private val returnType: TypeInformation[Row])
+  extends RichFlatMapFunction[Row, Row]
+  with ResultTypeQueryable[Row] {
+
+  Preconditions.checkNotNull(aggregatesLength)
+
+  private var intermediateRow: Row = _
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggFlatMapFunction.scala]: can be removed because we are replicating the input record.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggFlatMapFunction.scala [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+  // add one field to store window start
+  private val intermediateRowArity: Int = groupingKeysLength + aggregatesLength + 1
+
+  override def open(config: Configuration) {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggFlatMapFunction.scala]: `open()` can be removed

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggFlatMapFunction.scala [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    private val windowSlide: Long,
+    @transient private val returnType: TypeInformation[Row])
+  extends RichFlatMapFunction[Row, Row]
+  with ResultTypeQueryable[Row] {
+
+  Preconditions.checkNotNull(aggregatesLength)
+
+  private var intermediateRow: Row = _
+  // add one field to store window start
+  private val intermediateRowArity: Int = groupingKeysLength + aggregatesLength + 1
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggFlatMapFunction.scala]: can be removed

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggFlatMapFunction.scala [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+  * window. It is used for non-partial aggregations.
+  *
+  * @param aggregatesLength number of aggregate functions
+  * @param groupingKeysLength number of grouping keys
+  * @param windowSize window size of the sliding window
+  * @param windowSlide window slide of the sliding window
+  * @param returnType return type of this function
+  */
+class DataSetSlideTimeWindowAggFlatMapFunction(
+    private val aggregatesLength: Int,
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggFlatMapFunction.scala]: `aggregatesLength` and `groupingKeysLength` can be removed

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala [hunk 19]
```diff
@@ -186,6 +200,130 @@ object AggregateUtil {
...
+      inputType: RelDataType,
+      isParserCaseSensitive: Boolean)
+    : FlatMapFunction[Row, Row] = {
+
+    val aggregates = transformToAggregateFunctions(
+      namedAggregates.map(_.getKey),
+      inputType,
+      needRetraction = false)._2
+
+    val mapReturnType: RowTypeInfo = createDataSetAggregateBufferDataType(
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: output type should be equal to input type.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala [hunk 20]
```diff
@@ -186,6 +200,130 @@ object AggregateUtil {
...
+    */
+  def createDataSetSlideWindowPrepareFlatMapFunction(
+      window: LogicalWindow,
+      namedAggregates: Seq[CalcitePair[AggregateCall, String]],
+      groupings: Array[Int],
+      inputType: RelDataType,
+      isParserCaseSensitive: Boolean)
+    : FlatMapFunction[Row, Row] = {
+
+    val aggregates = transformToAggregateFunctions(
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: not needed because input type = output type

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala [hunk 21]
```diff
@@ -186,6 +200,130 @@ object AggregateUtil {
...
+    *                                              ^                 ^
+    *                                              |                 |
+    *                                 sum(y) aggOffsetInRow = 4      window start for pane mapping
+    * }}}
+    *
+    * NOTE: this function is only used for time-based sliding windows on batch tables.
+    */
+  def createDataSetSlideWindowPrepareFlatMapFunction(
+      window: LogicalWindow,
+      namedAggregates: Seq[CalcitePair[AggregateCall, String]],
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: remove unnecessary parameters `namedAggregates` and `groupings`

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggReduceCombineFunction.scala [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+  * It is used for sliding on batch for time-windows.
+  *
+  * @param aggregates aggregate functions
+  * @param groupingKeysLength number of grouping keys
+  * @param timeFieldPos position of aligned time field
+  * @param windowSize window size of the sliding window
+  * @param windowSlide window slide of the sliding window
+  * @param returnType return type of this function
+  */
+class DataSetSlideTimeWindowAggReduceCombineFunction(
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSlideTimeWindowAggReduceCombineFunction.scala]: Can be merged with `DataSetSlideTimeWindowAggReduceGroupFunction` because both can only be used if input is combinable.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetTumbleTimeWindowAggReduceGroupFunction.scala [hunk 23]
```diff
@@ -112,12 +112,10 @@ class DataSetTumbleTimeWindowAggReduceGroupFunction(
     }
 
     // get final aggregate value and set to output.
-    aggregateMapping.foreach {
-      case (after, previous) => {
+    aggregateMapping.foreach { case (after, previous) =>
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetTumbleTimeWindowAggReduceGroupFunction.scala]: revert this change? Will be fixed with #3489

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/IncrementalAggregateAllWindowFunction.scala [hunk 24]
```diff
@@ -53,7 +53,12 @@ class IncrementalAggregateAllWindowFunction[W <: Window](
 
     if (iterator.hasNext) {
       val record = iterator.next()
-      out.collect(record)
+      var i = 0
+      while (i < record.getArity) {
+        output.setField(i, record.getField(0))
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/IncrementalAggregateAllWindowFunction.scala]: `record.getField(0)` -> `record.getField(i)`?

Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/IncrementalAggregateAllWindowFunction.scala]: Good point!
